File: ios/SpentyAI/SpentyAI/Features/Vendors/VendorRepository.swift
import Foundation

// MARK: - Models

struct VendorBill: Codable, Identifiable {
    let id: String
    let billNumber: String?
    let date: String?
    let dueDate: String?
    let total: Double?
    let amountPaid: Double?
    let status: String?

    enum CodingKeys: String, CodingKey {
        case id, date, total, status
        case billNumber
        case dueDate
        case amountPaid
    }
}

struct CreateVendorRequest: Codable {
    let name: String
    let email: String?
    let phone: String?
    let gstin: String?
    let address: String?
}

// MARK: - Repository

final class VendorRepository {

    static let shared = VendorRepository()
    private init() {}

    // MARK: - CRUD

    func getVendors() async throws -> [Vendor] {
        try await APIClient.shared.get(APIEndpoints.vendors)
    }

    func getVendor(id: String) async throws -> Vendor {
        try await APIClient.shared.get(APIEndpoints.vendor(id))
    }

    func createVendor(_ request: CreateVendorRequest) async throws -> Vendor {
        try await APIClient.shared.post(APIEndpoints.vendors, body: request)
    }

    func updateVendor(id: String, _ request: CreateVendorRequest) async throws -> Vendor {
        try await APIClient.shared.put(APIEndpoints.vendor(id), body: request)
    }

    func deleteVendor(id: String) async throws {
        let _: EmptyResponse = try await APIClient.shared.delete(APIEndpoints.vendor(id))
    }

    // MARK: - Bills

    func getVendorBills(vendorId: String) async throws -> [VendorBill] {
        try await APIClient.shared.get(APIEndpoints.vendorBills(vendorId))
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/vendors/VendorRepository.kt
package com.spentyai.app.features.vendors

import com.spentyai.app.core.models.Bill
import com.spentyai.app.core.models.Vendor
import com.spentyai.app.core.network.ApiClient
import com.spentyai.app.core.network.ApiResult
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

class VendorRepository(private val apiClient: ApiClient) {

    // CRUD

    suspend fun fetchAll(): ApiResult<List<Vendor>> =
        apiClient.safeApiCall { apiClient.endpoints.getVendors() }

    suspend fun fetch(id: String): ApiResult<Vendor> =
        apiClient.safeApiCall { apiClient.endpoints.getVendor(id) }

    suspend fun create(payload: JsonObject): ApiResult<Vendor> =
        apiClient.safeApiCall { apiClient.endpoints.createVendor(payload) }

    suspend fun update(id: String, payload: JsonObject): ApiResult<Vendor> =
        apiClient.safeApiCall { apiClient.endpoints.updateVendor(id, payload) }

    suspend fun delete(id: String): ApiResult<JsonObject> =
        apiClient.safeApiCall { apiClient.endpoints.deleteVendor(id) }

    // Bills for a vendor
    suspend fun fetchVendorBills(vendorId: String): ApiResult<List<Bill>> =
        apiClient.safeApiCall { apiClient.endpoints.getBills(vendorId = vendorId) }
}

fun buildVendorPayload(
    name: String,
    email: String?,
    phone: String?,
    gstin: String?,
    address: String?,
    notes: String?
): JsonObject = buildJsonObject {
    put("name", name)
    email?.let { put("email", it) }
    phone?.let { put("phone", it) }
    gstin?.let { put("tax_id", it) }
    address?.let {
        put("address", buildJsonObject {
            put("street", it)
        })
    }
    notes?.let { put("notes", it) }
}
